core/pedersen_hash.go

```diff
@@ -0,0 +1,45 @@
+package core
+
+import (
+	"errors"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+// PedersenArray implements [Pedersen array hashing]
+//
+// [Pedersen array hashing]: https://docs.starknet.io/documentation/develop/Hashing/hash-functions/#pedersen_hash
+func PedersenArray(elems ...*felt.Felt) (*felt.Felt, error) {
+	if len(elems) < 3 {
+		return nil, errors.New("number of elems must be more than 2")
+	}
+
+	var err error
+	d := new(felt.Felt).SetZero()
```

Comment:
Does `new` always allocate on heap? Or does it do escape analysis?

`d` (same goes for `l`)  never escapes this function scope, so there is no need to allocate in on the heap. We should make sure that any ephemeral variable is allocated on stack.

Reply:
[Looks like](https://stackoverflow.com/questions/67895905/go-memory-allocation-new-objects-pointers-and-escape-analysis) that the `new` keyword doesn't necessarily allocate on the stack.

Reply:
[This](https://blog.devgenius.io/in-depth-analysis-of-golang-memory-escape-edfbfb856913) article mentions the allocation happens after escape analysis.